src/OctoshiftCLI.Tests/bbs2gh/Services/BbsArchiveDownloaderTests.cs

```diff
@@ -12,72 +13,82 @@ public sealed class BbsArchiveDownloaderTests : IDisposable
 {
     private const int EXPORT_JOB_ID = 1;
     private const string BBS_HOME_DIRECTORY = "BBS_HOME";
+    private const string TARGET_DIRECTORY = "TARGET";
 
     private readonly string _exportArchiveFilename = $"Bitbucket_export_{EXPORT_JOB_ID}.tar";
     private readonly Mock<OctoLogger> _mockOctoLogger = TestHelpers.CreateMock<OctoLogger>();
     private readonly Mock<FileSystemProvider> _mockFileSystemProvider = TestHelpers.CreateMock<FileSystemProvider>();
+    private readonly Mock<ISftpClient> _mockSftpClient = new();
     private readonly BbsArchiveDownloader _bbsArchiveDownloader;
-    private readonly ScpClient _scpClient;
 
     public BbsArchiveDownloaderTests()
     {
-        _scpClient = new ScpClient("host", "username", "password");
-        _bbsArchiveDownloader = new BbsArchiveDownloader(_mockOctoLogger.Object, _scpClient)
+        _bbsArchiveDownloader = new BbsArchiveDownloader(_mockOctoLogger.Object, _mockSftpClient.Object)
         {
             FileSystemProvider = _mockFileSystemProvider.Object,
             BbsSharedHomeDirectory = BBS_HOME_DIRECTORY
         };
+
+        _mockSftpClient.Setup(m => m.Exists(It.IsAny<string>())).Returns(true);
+
+        var mockAsyncResult = new Mock<IAsyncResult>();
+        mockAsyncResult.Setup(m => m.IsCompleted).Returns(true);
+        _mockSftpClient
+            .Setup(m => m.BeginDownloadFile(It.IsAny<string>(), It.IsAny<Stream>(), It.IsAny<AsyncCallback>(), It.IsAny<object>(), It.IsAny<Action<ulong>>()))
+            .Returns(mockAsyncResult.Object);
     }
 
     [Fact]
-    public void Download_Calls_ScpClient_Download_With_Correct_Params()
+    public async Task Download_Calls_SftpClinet_DownloadFile_With_Correct_Params()
     {
         // Arrange
         var expectedSourceArchiveFullName = Path.Combine(BBS_HOME_DIRECTORY, "data/migration/export", _exportArchiveFilename);
-
-        string actualSource = null;
-        string actualTarget = null;
-        _bbsArchiveDownloader.ScpDownload = (source, target) =>
-        {
-            actualSource = source;
-            actualTarget = target.Name;
-        };
+        var expectedTargetArchiveFullName = Path.Combine(TARGET_DIRECTORY, _exportArchiveFilename);
 
         // Act
-        _bbsArchiveDownloader.Download(EXPORT_JOB_ID);
+        await _bbsArchiveDownloader.Download(EXPORT_JOB_ID, TARGET_DIRECTORY);
 
         // Assert
-        actualSource.Should().Be(expectedSourceArchiveFullName);
-        actualTarget.Should().Be(_exportArchiveFilename);
+        _mockSftpClient.Verify(m =>
+            m.BeginDownloadFile(
+                It.Is<string>(actual => actual == expectedSourceArchiveFullName),
+                It.IsAny<Stream>(),
```

Comment:
Wouldn't it be better to do a _mockFileSystemProvider.Setup(...) that returns some stream, then verify that stream is what's passed into BeginDownloadFile? If we did that we wouldn't need the Verify on _mockFileSystemProvider either.

Reply:
I tried to do it in the previous commit, but unfortunately `FileStream` is hard and even impossible to mock and creating a dummy one is also impossible. In the previous commit I tried to mock the `FileStream` like the following:
```c#
var expectedTargetArchive = new Mock<FileStream>(new SafeFileHandle(IntPtr.Zero, false), FileAccess.Read);
```
but it failed our `Windows` CI because of the fact that `.NET 6` added some extra checks for `SafeFileHandle` so it's no longer possible to create invalid file handles (eg. `new SafeFileHandle(IntPtr.Zero, false)`).
The only way to test/mock `FileStream` (or any other `Stream`) would be to create abstractions (eg. `IFileStream`), but IMO it's going to be too much effort for not that much of a gain!